File: client/lib/src/directory/root_quorum.dart
import 'dart:typed_data';

import 'package:sodium/sodium.dart';

import '../crypto/ed25519.dart';
import 'directory_record.dart';
import 'directory_rejection.dart';

/// Verifies that a set of root signatures meets the quorum threshold.
class RootQuorum {
  RootQuorum._();

  /// Verifies [sigs] over [signedBytes] at [issuedAtMs] against [pinned].
  ///
  /// - Duplicate key IDs count as a single valid signature.
  /// - Every signature must reference a known, in-window root key.
  /// - Throws a subtype of [DirectoryRejection] on any failure.
  static void verify(
    Sodium sodium,
    Uint8List signedBytes,
    List<RootSignature> sigs,
    int issuedAtMs,
    PinnedRoot pinned,
  ) {
    final validKeyIds = <String>{};

    for (final sig in sigs) {
      final root = _findRoot(sig.rootKeyId, pinned);
      if (root == null) {
        throw RootKeyNotFound(
            'Root key not found: ${_hex(sig.rootKeyId)}');
      }

      if (issuedAtMs < root.validFrom || issuedAtMs >= root.validUntil) {
        throw RootKeyExpired(
            'Root key outside valid window at issued_at=$issuedAtMs');
      }

      if (!Ed25519.verifyDetached(
          sodium, root.publicKey, signedBytes, sig.signature)) {
        throw SignatureInvalid(
            'Signature invalid for root key: ${_hex(sig.rootKeyId)}');
      }

      validKeyIds.add(_hex(sig.rootKeyId));
    }

    if (validKeyIds.length < pinned.threshold) {
      throw QuorumBelowThreshold(validKeyIds.length, pinned.threshold);
    }
  }

  static RootEntry? _findRoot(Uint8List rootKeyId, PinnedRoot pinned) {
    for (final entry in pinned.roots) {
      if (_bytesEqual(entry.rootKeyId, rootKeyId)) return entry;
    }
    return null;
  }

  static bool _bytesEqual(Uint8List a, Uint8List b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  static String _hex(Uint8List bytes) {
    return bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  }
}

File: src/main/java/com/erikromson/datawallet/directory/RootQuorum.java
package com.erikromson.datawallet.directory;

import com.erikromson.datawallet.crypto.Ed25519;

import java.util.Arrays;
import java.util.HashSet;
import java.util.HexFormat;
import java.util.List;
import java.util.Set;

public final class RootQuorum {

    private RootQuorum() {}

    public static void verify(byte[] signedBytes, List<RootSignature> sigs, long issuedAtMs, PinnedRoot pinned) {
        Set<String> validKeyIds = new HashSet<>();

        for (RootSignature sig : sigs) {
            PinnedRoot.RootEntry root = findRoot(sig.rootKeyId(), pinned);
            if (root == null) {
                throw new DirectoryRejection.RootKeyNotFound(
                        "Root key not found: " + HexFormat.of().formatHex(sig.rootKeyId()));
            }

            if (issuedAtMs < root.validFrom() || issuedAtMs >= root.validUntil()) {
                throw new DirectoryRejection.RootKeyExpired(
                        "Root key outside valid window at issued_at=" + issuedAtMs);
            }

            if (!Ed25519.verifyDetached(root.publicKey(), signedBytes, sig.signature())) {
                throw new DirectoryRejection.SignatureInvalid(
                        "Signature invalid for root key: " + HexFormat.of().formatHex(sig.rootKeyId()));
            }

            validKeyIds.add(HexFormat.of().formatHex(sig.rootKeyId()));
        }

        if (validKeyIds.size() < pinned.threshold()) {
            throw new DirectoryRejection.QuorumBelowThreshold(validKeyIds.size(), pinned.threshold());
        }
    }

    private static PinnedRoot.RootEntry findRoot(byte[] rootKeyId, PinnedRoot pinned) {
        for (PinnedRoot.RootEntry entry : pinned.roots()) {
            if (Arrays.equals(entry.rootKeyId(), rootKeyId)) {
                return entry;
            }
        }
        return null;
    }
}
